pkg/providers/generic.go

```diff
@@ -0,0 +1,103 @@
+package providers
+
+import (
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/apex/log"
+	"github.com/marcosnils/bin/pkg/assets"
+)
+
+type generic struct {
+	baseURL   string
+	name      string
+	version   string
+	os        string
+	arch      string
+	ext       string
+	latestURL string
+}
+
+func (g *generic) GetID() string {
+	return "generic"
+}
+
+func (g *generic) GetLatestVersion() (string, string, error) {
+	log.Debugf("Getting latest release for %s", g.name)
+
+	resp, err := http.Get(g.latestURL)
+	if err != nil {
+		return "", "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", "", fmt.Errorf("HTTP response error: %v", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", "", fmt.Errorf("Error reading response body: %v", err)
+	}
+
+	version := strings.TrimSuffix(string(body), "\n")
+	_, link := g.fmtNameLink(version)
+
+	return version, link, nil
+}
+
+func (g *generic) Fetch(opts *FetchOpts) (*File, error) {
+	var version string
+	if len(g.version) > 0 {
+		log.Infof("Getting release %s for %s", g.version, g.name)
+		version = g.version
+	} else {
+		latest, _, err := g.GetLatestVersion()
+		if err != nil {
+			return nil, err
+		}
+		version = latest
+	}
+
+	name, link := g.fmtNameLink(version)
+	candidates := []*assets.Asset{{Name: name, URL: link}}
+
+	f := assets.NewFilter(&assets.FilterOpts{SkipScoring: opts.All, PackagePath: opts.PackagePath, SkipPathCheck: opts.SkipPatchCheck})
+	gf, err := f.FilterAssets(g.name, candidates)
+	if err != nil {
+		return nil, err
+	}
+
+	outFile, err := f.ProcessURL(gf)
+	if err != nil {
+		return nil, err
+	}
+
+	return &File{Data: outFile.Source, Name: assets.SanitizeName(outFile.Name, version), Hash: sha256.New(), Version: version}, nil
+}
+
+func (g *generic) fmtNameLink(version string) (string, string) {
+	fname := fmt.Sprintf("%s-%s-%s-%s.%s", g.name, version, g.os, g.arch, g.ext)
+	return fname, fmt.Sprintf("%s/%s", g.baseURL, fname)
+}
+
+func newGeneric(u *url.URL, latestURL string) (Provider, error) {
```

Comment:
seems like this provider is not very generic since it expects the URL to have a specific format or otherwise it'll fail. I'll leave a more general comment about this in the issue.